File: godbrain_core/tools/cs2_market_crawler.py
import asyncio
import logging
import time
import urllib.request
import urllib.parse
import json
from pymongo import MongoClient

logging.basicConfig(level=logging.INFO, format="%(asctime)s - [CS2_MARKET_CRAWLER] - %(message)s")
logger = logging.getLogger("CS2Market")

client = MongoClient('mongodb://localhost:27017/')
db = client['godbrain']
price_cache = db['cs2_price_cache']

# Strict rate limit: 1 request every 6 seconds = 10 per minute = 600 per hour.
# This prevents Steam from throwing a 429 Too Many Requests ban.
RATE_LIMIT_DELAY = 6.0 

TARGET_CATEGORIES = [
    "AK-47", "M4A4", "M4A1-S", "AWP", "Desert Eagle", "Karambit", "Butterfly Knife", "M9 Bayonet"
]

COMMON_SKINS = [
    "AK-47 | Redline (Field-Tested)",
    "M4A1-S | Printstream (Field-Tested)",
    "AWP | Asiimov (Field-Tested)",
    "Desert Eagle | Printstream (Field-Tested)",
    "AK-47 | Slate (Field-Tested)"
]

class CS2MarketCrawler:
    def __init__(self):
        logger.info("Initializing CS2 Rate-Limited Market Crawler...")

    def fetch_price(self, market_hash_name):
        """Fetches the price from Steam with strict starvation to avoid 429s."""
        encoded_name = urllib.parse.quote(market_hash_name)
        url = f"https://steamcommunity.com/market/priceoverview/?appid=730&currency=1&market_hash_name={encoded_name}"
        
        headers = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
            'Accept-Language': 'en-US,en;q=0.9'
        }
        
        try:
            req = urllib.request.Request(url, headers=headers)
            response = urllib.request.urlopen(req)
            data = json.loads(response.read().decode('utf-8'))
            
            if data and data.get('success'):
                price_info = {
                    "market_hash_name": market_hash_name,
                    "lowest_price": data.get('lowest_price', 'N/A'),
                    "volume": data.get('volume', 'N/A'),
                    "timestamp": time.time()
                }
                # Upsert into MongoDB
                price_cache.update_one(
                    {"market_hash_name": market_hash_name},
                    {"$set": price_info},
                    upsert=True
                )
                logger.info(f"[CACHE UPDATED] {market_hash_name}: {price_info['lowest_price']} (Volume: {price_info['volume']})")
                return True
            else:
                logger.warning(f"Steam returned success=false for {market_hash_name}")
                return False
                
        except urllib.error.HTTPError as e:
            logger.error(f"HTTP Error {e.code} for {market_hash_name}. Rate limit hit?")
            return False
        except Exception as e:
            logger.error(f"Error fetching {market_hash_name}: {e}")
            return False

    async def run_crawler_loop(self):
        """Infinite loop that slowly trickles requests to build the database."""
        logger.info(f"Starting crawler loop. Strict delay set to {RATE_LIMIT_DELAY}s between requests.")
        
        while True:
            for skin in COMMON_SKINS:
                # Check if we already fetched it recently (within 12 hours)
                cached = price_cache.find_one({"market_hash_name": skin})
                if cached and (time.time() - cached.get('timestamp', 0)) < (12 * 3600):
                    logger.debug(f"Skipping {skin}, cache is fresh.")
                    continue
                
                logger.info(f"Fetching live data for: {skin}")
                self.fetch_price(skin)
                
                # The Golden Rule: Starve the API to stay under the radar
                await asyncio.sleep(RATE_LIMIT_DELAY)
            
            logger.info("Cycle complete. Resting for 1 hour before next cache refresh...")
            await asyncio.sleep(3600)

if __name__ == "__main__":
    crawler = CS2MarketCrawler()
    asyncio.run(crawler.run_crawler_loop())

File: godbrain_core/cpp_tools/cs2_market_crawler.cpp
#define WIN32_LEAN_AND_MEAN
#include <windows.h>
#include <iostream>
#include <string>
#include <vector>
#include <thread>
#include <chrono>

#include "../cpp_kernel/json.hpp"
using json = nlohmann::json;

const double RATE_LIMIT_DELAY = 6.0;

std::string exec_cmd(const std::string& cmd) {
    char buffer[256];
    std::string result = "";
    FILE* pipe = _popen(cmd.c_str(), "r");
    if (!pipe) return "";
    while (fgets(buffer, sizeof(buffer), pipe) != NULL) {
        result += buffer;
    }
    _pclose(pipe);
    return result;
}

std::string url_encode(const std::string &value) {
    std::string escaped;
    for (char c : value) {
        if (isalnum((unsigned char)c) || c == '-' || c == '_' || c == '.' || c == '~') {
            escaped += c;
        } else {
            char buf[10];
            snprintf(buf, sizeof(buf), "%%%02X", (unsigned char)c);
            escaped += buf;
        }
    }
    return escaped;
}

std::vector<std::string> get_targets() {
    return {
        "AK-47 | Redline (Field-Tested)", "AK-47 | Slate (Field-Tested)"
    };
}

void fetch_price(const std::string& market_hash_name) {
    std::string url = "https://steamcommunity.com/market/priceoverview/?appid=730&currency=1&market_hash_name=" + url_encode(market_hash_name);
    std::string cmd = "curl -s -A \"Mozilla/5.0 (Windows NT 10.0; Win64; x64)\" \"" + url + "\"";
    
    std::string out = exec_cmd(cmd);
    
    try {
        json data = json::parse(out);
        if (data.value("success", false)) {
            std::string lowest_price = data.value("lowest_price", "N/A");
            std::string volume = data.value("volume", "N/A");
            std::cout << "[CACHE UPDATED] " << market_hash_name << ": " << lowest_price << " (Vol: " << volume << ")\n";
        } else {
            std::cout << "[WARNING] Steam success=false for " << market_hash_name << "\n";
        }
    } catch (...) {
        std::cout << "[ERROR] JSON parse error for " << market_hash_name << "\n";
    }
}

int main(int argc, char* argv[]) {
    std::cout << "[CS2_MARKET_CRAWLER] Initializing Expanded CS2 Market Crawler (C++ Native)...\n";
    auto targets = get_targets();
    std::cout << "Loaded " << targets.size() << " priority targets. Delay set to " << RATE_LIMIT_DELAY << "s.\n";

    bool daemon_mode = (argc > 1 && std::string(argv[1]) == "--daemon");

    do {
        for (const auto& skin : targets) {
            // In a full implementation we'd check a MongoDB cache here
            fetch_price(skin);
            std::this_thread::sleep_for(std::chrono::milliseconds(static_cast<int>(RATE_LIMIT_DELAY * 1000)));
        }
        
        if (daemon_mode) {
            std::cout << "Main loadout cycle complete. Sleeping 1 hour.\n";
            std::this_thread::sleep_for(std::chrono::hours(1));
        }
    } while (daemon_mode);

    return 0;
}